file.c

```diff
@@ -3596,34 +3631,298 @@ rb_home_dir_of(VALUE user, VALUE result)
-		return result;
-	    }
-	    endpwent();
-	    rb_raise(rb_eArgError, "couldn't find HOME for login `%s' -- expanding `~'",
-		     login);
-	}
-	else {
-	    rb_raise(rb_eArgError, "couldn't find login name -- expanding `~'");
-	}
+    /* We'll look up the user's default home dir in the password db by login
+     * name, if possible, and failing that will fall back to looking the
+     * information up by uid (as would be needed for processes that are not a
+     * descendant of login(1) or a work-alike).
+     *
+     * While the lookup by uid is more likely to succeed (since we always have
+     * a uid, but may or may not have a login name), we prefer first looking
+     * up by name to accommodate the possibility of multiple login names (each
+     * with its own record in the password database, so each with a
+     * potentially different home directory) being mapped to the same uid (as
+     * explicitly allowed for by POSIX; see getlogin(3posix)).
+     */
+
+    VALUE pw_dir_str_nm = rb_getpwdirnam_for_login();
+    if (!NIL_P(pw_dir_str_nm)) {
+        /* found it */
+        copy_home_path(result, RSTRING_PTR(pw_dir_str_nm));
+        rb_str_resize(pw_dir_str_nm, 0);
+        return result;
+    }
+
+    VALUE pw_dir_str_uid = rb_getpwdiruid();
+    if (!NIL_P(pw_dir_str_uid)) {
+        /* found it */
+        copy_home_path(result, RSTRING_PTR(pw_dir_str_uid));
+        rb_str_resize(pw_dir_str_uid, 0);
+        return result;
+    }
+
+    rb_raise(rb_eArgError, "couldn't find home for uid `%ld'", (long)getpid());
+}
+
+/**
+ * Best-effort attempt to obtain the name of the login user, if any,
+ * associated with the process. Processes not descended from login(1) (or
+ * similar) may not have a logged-in user; returns Qnil in that case.
+ */
+VALUE
+rb_getlogin(void)
+{
+#if ( !defined(USE_GETLOGIN_R) && !defined(USE_GETLOGIN) )
+    return Qnil;
+#else
+    char MAYBE_UNUSED(*login) = NULL;
+
+# ifdef USE_GETLOGIN_R
+
+    long loginsize = GETLOGIN_R_SIZE_INIT;  /* maybe -1 */
+
+    if (loginsize < 0)
+        loginsize = GETLOGIN_R_SIZE_DEFAULT;
+
+    VALUE getlogin_tmp = rb_str_tmp_new(loginsize);
+
+    login = RSTRING_PTR(getlogin_tmp);
+    loginsize = rb_str_capacity(getlogin_tmp);
+    rb_str_set_len(getlogin_tmp, loginsize);
+
+    int gle;
+    errno = 0;
+    while ((gle = getlogin_r(login, loginsize)) != 0) {
+
+        if (gle == ENOTTY || gle == ENXIO || ENOENT) {
+            rb_str_resize(getlogin_tmp, 0);
+            return Qnil;
+        }
+
+        if (gle != ERANGE || loginsize >= GETLOGIN_R_SIZE_LIMIT) {
+            rb_str_resize(getlogin_tmp, 0);
+            rb_syserr_fail(gle, "getlogin_r");
+        }
+
+        rb_str_modify_expand(getlogin_tmp, loginsize);
+        login = RSTRING_PTR(getlogin_tmp);
+        loginsize = rb_str_capacity(getlogin_tmp);
+    }
+
+    if (login == NULL) {
+        rb_str_resize(getlogin_tmp, 0);
+        return Qnil;
+    }
+
+    VALUE result = rb_str_new_cstr(login);
```

Comment:
No need to create a new `String` from `login`, it is a pointer into a `String`.

Reply:
Roger that; thanks.